master/internal/spec_util.go

```diff
@@ -0,0 +1,135 @@
+package internal
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/pkg/errors"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+	k8sV1 "k8s.io/api/core/v1"
+
+	"github.com/determined-ai/determined/master/internal/config"
+	"github.com/determined-ai/determined/master/internal/grpcutil"
+	"github.com/determined-ai/determined/master/internal/sproto"
+	"github.com/determined-ai/determined/master/internal/user"
+	"github.com/determined-ai/determined/master/pkg/archive"
+	pkgCommand "github.com/determined-ai/determined/master/pkg/command"
+	"github.com/determined-ai/determined/master/pkg/model"
+	"github.com/determined-ai/determined/master/pkg/schemas"
+	"github.com/determined-ai/determined/master/pkg/schemas/expconf"
+	"github.com/determined-ai/determined/master/pkg/tasks"
+	"github.com/determined-ai/determined/proto/pkg/utilv1"
+)
+
+// ResolveResources - Validate ResoucePool and check for availability.
+func (m *Master) ResolveResources(
+	resourcePool string,
+	slots int,
+	workspaceID int,
+) (string, []pkgCommand.LaunchWarning, error) {
+	poolName, err := m.rm.ResolveResourcePool(
+		resourcePool, workspaceID, slots)
+	if err != nil {
+		return "", nil, status.Errorf(codes.InvalidArgument, err.Error())
+	}
+	if err = m.rm.ValidateResources(poolName, slots, true); err != nil {
+		return "nil", nil, fmt.Errorf("validating resources: %v", err)
+	}
+
+	launchWarnings, err := m.rm.ValidateResourcePoolAvailability(
+		&sproto.ValidateResourcePoolAvailabilityRequest{
+			Name:  poolName,
+			Slots: slots,
+		},
+	)
+	if err != nil {
+		return "", launchWarnings, fmt.Errorf("checking resource availability: %v", err.Error())
+	}
+	if m.config.ResourceManager.AgentRM != nil &&
+		m.config.LaunchError &&
+		len(launchWarnings) > 0 {
+		return "", nil, errors.New("slots requested exceeds cluster capacity")
+	}
+
+	return poolName, launchWarnings, nil
+}
+
+// Fill and return TaskSpec.
+func (m *Master) fillTaskSpec(
+	poolName string,
+	agentUserGroup *model.AgentUserGroup,
+	userModel *model.User,
+) (tasks.TaskSpec, error) {
+	taskContainerDefaults, err := m.rm.TaskContainerDefaults(
+		poolName,
+		m.config.TaskContainerDefaults,
+	)
+	if err != nil {
+		return tasks.TaskSpec{}, fmt.Errorf("getting TaskContainerDefaults: %v", err)
+	}
+	taskSpec := *m.taskSpec
+	taskSpec.TaskContainerDefaults = taskContainerDefaults
+	taskSpec.AgentUserGroup = agentUserGroup
+	taskSpec.Owner = userModel
+	return taskSpec, nil
+}
+
+func fillTaskConfig(slots int, taskSpec tasks.TaskSpec, environment *model.Environment) {
```

Comment:
maybe fill pod spec instead? also maybe a mini-convention like calling `environment` `out`/`res` since it is mutated would be nice

this one definitely could be like this instead, but i don't have super strong feelings
```suggestion
func fillPodSpec(slots int, taskSpec tasks.TaskSpec) *k8sv1.Pod {
```